### PR TITLE
ID-43 - messy WIP towards stubbing Stripe for customer create and update

### PR DESCRIPTION
**Pull request by NoelLH**

### CODE REVIEW

File: src/Client/Stripe.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+declare(strict_types=1);
+
+namespace BigGive\Identity\Client;
+
+use BigGive\Identity\Tests\Client\Stripe\StubCustomerService;
+use Stripe\StripeClient;
+
+class Stripe
```
Comment (bdsl) [inline src/Client/Stripe.php:10]:
```suggestion
/**
 * @property ?StubCustomerService customers
 */
class Stripe
```

File: src/Client/Stripe.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            $this->stripeNativeClient = new StripeClient($stripeOptions);
+            return;
+        }
+
+        // Else, fake everything we must for load tests. Anything not stubbed explicitly will crash in
+        // stub mode.
+        $this->customers = new StubCustomerService();
+    }
+
+    public function __get($name)
```
Comment (bdsl) [inline src/Client/Stripe.php:30]:
```suggestion
    public function __get(string $name)
```

File: tests/Client/Stripe/StubCustomerService.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        );
+        /** @var array $customer */
+        $customer = json_decode($customerJSON, true, 512, JSON_THROW_ON_ERROR);
+        $customer['id'] = 'cus_' . bin2hex(random_bytes(12));
+
+        $customer = Customer::constructFrom($customer);
+
+        \assert($customer instanceof Customer);
+
+        return $customer;
```
Comment (bdsl) [inline tests/Client/Stripe/StubCustomerService.php:25]:
```suggestion
        return $customer Customer::constructFrom($customer);
```

Psalm says the assert here is redundant.

**Review by bdsl — APPROVED** [review]
Looks good to me considering the urgency. I've added suggestions that fix two of the Psalm errors, maybe we can add the rest to the baseline and then revert this PR after load tests are done or suppress them inline.

If we want to keep working on a custom made wrapper for Stripe longer term I'd say we probably don't need to mirror the actual stripe API quite so precisely, we can make something with better DX, and also do polymorphism to move the `if ($stubbed)` part to dependancies.php

File: src/Client/Stripe.php (hunk 2 above)
Comment (bdsl) [reply src/Client/Stripe.php:30]: And of course let's not keep using `__get` long term here - makes static analysis pretty much impossible, and the fact that its there means Psalm won't alert us if we have a typo like `$this->stripeClient->ccustomers`

File: src/Client/Stripe.php (hunk 1 above)
Comment (bdsl) [inline src/Client/Stripe.php:10]:
```suggestion
use BigGive\Identity\Tests\Client\Stripe\StubCustomerService;
use Stripe\Service\CustomerService;
use Stripe\StripeClient;

/**
 * @property ?StubCustomerService|CustomerService customers
 */
class Stripe
```

File: tests/Client/Stripe/StubCustomerService.php (hunk 3 above)
Comment (NoelLH) [reply tests/Client/Stripe/StubCustomerService.php:25]: Makes sense, addressed locally as I think the suggestion has a typo

**Pull request merged**